Release Wasm 2.0.0

rossberg opened the pull request:
Release package for Wasm 2.0 interpreter

mseri commented on packages/wasm/wasm.2.0.0/opam:
```diff
@@ -0,0 +1,23 @@
+  "ocamlbuild" {build}
+  "dune" {>= "2.9"}
+]
+synopsis:
```
```suggestion
available: arch != "arm32" & arch != "ppc64"
```

Due to:
```
#=== ERROR while compiling wasm.2.0.0 =========================================#
# context              2.1.2 | linux/ppc64 | ocaml-base-compiler.4.14.0 | pinned(https://github.com/WebAssembly/spec/archive/opam-2.0.0.zip)
# path                 ~/.opam/4.14/.opam-switch/build/wasm.2.0.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build make -C interpreter opt unopt
# exit-code            2
# env-file             ~/.opam/log/wasm-3246-4976a9.env
# output-file          ~/.opam/log/wasm-3246-4976a9.out
### output ###
# make: Entering directory '/home/opam/.opam/4.14/.opam-switch/build/wasm.2.0.0/interpreter'
# echo >_tags "true: bin_annot"
# echo >>_tags "true: debug"
# echo >>_tags "<{util,syntax,binary,text,valid,runtime,exec,script,host,main,tests}/*.cmx>: for-pack(Wasm)"
# ocamlbuild -lexflags -ml -cflags '-w +a-4-27-42-44-45-70 -warn-error +a-3' -I util -I syntax -I binary -I text -I valid -I runtime -I exec -I script -I host -I main -I tests -libs bigarray -quiet main.native
# + /home/opam/.opam/4.14/bin/ocamlopt.opt -c -w +a-4-27-42-44-45-70 -warn-error +a-3 -g -bin-annot -for-pack Wasm -I text -I valid -I util -I tests -I syntax -I script -I runtime -I main -I host -I exec -I binary -o text/lexer.cmx text/lexer.ml
# /opam-tmp/camlasm11a009.s: Assembler messages:
# /opam-tmp/camlasm11a009.s:97170: Error: operand out of range (0x00000000000080ec is not between 0xffffffffffff8000 and 0x0000000000007ffc)
# /opam-tmp/camlasm11a009.s:97172: Error: operand out of range (0x00000000000080d8 is not between 0xffffffffffff8000 and 0x0000000000007ffc)
# /opam-tmp/camlasm11a009.s:97231: Error: operand out of range (0x000000000000800c is not between 0xffffffffffff8000 and 0x0000000000007ffc)
# /opam-tmp/camlasm11a009.s:97235: Error: operand out of range (0x0000000000008000 is not between 0xffffffffffff8000 and 0x0000000000007ffc)
# File "text/lexer.ml", line 1:
# Error: Assembler error, input left in file /opam-tmp/camlasm11a009.s
# Command exited with code 2.
# make: *** [Makefile:70: main.native] Error 10
# rm _tags
# make: Leaving directory '/home/opam/.opam/4.14/.opam-switch/build/wasm.2.0.0/interpreter'
```
and
```
#=== ERROR while compiling wasm.2.0.0 =========================================#
# context              2.1.2 | linux/arm32 | ocaml-base-compiler.4.14.0 | pinned(https://github.com/WebAssembly/spec/archive/opam-2.0.0.zip)
# path                 ~/.opam/4.14/.opam-switch/build/wasm.2.0.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build make -C interpreter opt unopt
# exit-code            2
# env-file             ~/.opam/log/wasm-2692-cf35b8.env
# output-file          ~/.opam/log/wasm-2692-cf35b8.out
### output ###
# make: Entering directory '/home/opam/.opam/4.14/.opam-switch/build/wasm.2.0.0/interpreter'
# echo >_tags "true: bin_annot"
# echo >>_tags "true: debug"
# echo >>_tags "<{util,syntax,binary,text,valid,runtime,exec,script,host,main,tests}/*.cmx>: for-pack(Wasm)"
# ocamlbuild -lexflags -ml -cflags '-w +a-4-27-42-44-45-70 -warn-error +a-3' -I util -I syntax -I binary -I text -I valid -I runtime -I exec -I script -I host -I main -I tests -libs bigarray -quiet main.native
# + /home/opam/.opam/4.14/bin/ocamlopt.opt -c -w +a-4-27-42-44-45-70 -warn-error +a-3 -g -bin-annot -for-pack Wasm -I text -I valid -I util -I tests -I syntax -I script -I runtime -I main -I host -I exec -I binary -o text/lexer.cmx text/lexer.ml
# /opam-tmp/camlasmd158c0.s: Assembler messages:
# /opam-tmp/camlasmd158c0.s:47874: Error: value of 102622 too large for field of 2 bytes at 79974
# /opam-tmp/camlasmd158c0.s:47875: Error: value of 102622 too large for field of 2 bytes at 79976
# /opam-tmp/camlasmd158c0.s:47877: Error: value of 102491 too large for field of 2 bytes at 79980
[...]
# /opam-tmp/camlasmd158c0.s:48531: Error: value of 102217 too large for field of 2 bytes at 81174
# File "text/lexer.ml", line 1:
# Error: Assembler error, input left in file /opam-tmp/camlasmd158c0.s
# Command exited with code 2.
# make: *** [Makefile:70: main.native] Error 10
```

rossberg replied:
Yes, this is rather annoying. Have a PR [in flight](https://github.com/WebAssembly/spec/pull/1482) to refactor the lexer to avoid the table overflow. Any idea why this only is an issue on these two architectures?

mseri replied:
No idea. But only the tests are failing, you can keep them on and failing if you prefer (to warn potential users) or disable the architectures altogether. Here I am just suggesting to disable the tests

msprotz commented:
That's surprising. @R1kM did someone mutate the tag?

rossberg commented on packages/wasm/wasm.2.0.0/opam:
```diff
@@ -0,0 +1,23 @@
+  "ocaml" {>= "4.08.0"}
+  "ocamlfind" {build}
+  "ocamlbuild" {build}
+  "dune" {>= "2.9"}
```
Yeah, unfortunately, the CI linter complains otherwise (unless there is a way to shut it up?).

rossberg reviewed (commented):
@msprotz, is a change to Kremlin deps needed regardless?

FWIW, I don't understand much about Opam, but it seems dubious that it defaults to assuming compat even with future major versions – by definition, major version updates are expected to break stuff?

mseri commented:
> FWIW, I don't understand much about Opam, but it seems dubious that it defaults to assuming compat even with future major versions – by definition, major version updates are expected to break stuff?

many packages don’t follow semver, and the module system makes it tricky to really follow it.

Opam itself just uses the metadata in the opam files to compute the set of compatible dependencies. When a release breaks old packages, we simply update the metadata alongside. It is not a problem and will maintain the opam repository stable, in this case, once the tarball for Kremlin is fixed, if there is a failure we will just update the upper bound in the relevant opam file in the opam repository

mseri commented:
> can you tell us whether an update to kremlin is desirable

Yes. That is what we normally do

rossberg commented:
> if there is a failure we will just update the upper bound in the relevant opam file in the opam repository

Pardon my ignorance, but who is responsible for updating it, and does a failure like this block the upstream update?

mseri commented:
Us maintainers or who submits the PR. We usually help, but if you do it yourself alongside the PR this helps us dealing with it faster. In this case I would not make it block the PR since the package is not installable, instead I’d open an issue saying it is not installable due to modified tarball and move on (I am planning to do the tomorrow). Often we can recover the tarballs from some cache and fix it, but I cannot check before tomorrow

mseri commented:
@msprotz @R1kM a cached version of kremlin zip is at http://opam.ocaml.org/cache/md5/b8/b8bf9167faca0927f86cf5bb1d3d987a can you please upload it on the github release? Otherwise I can add it to the backup opam source aechives and point at those some time tomorrow: http://opam.ocaml.org/cache/md5/b8/b8bf9167faca0927f86cf5bb1d3d987a

rossberg commented:
Okay, I removed the dune dependency, disabled the failing tests, and updated Kremlin's dependency. However, I assume the latter will generate loads of errors for now, as long as Kremlin is not installable. PTAL

mseri commented:
> @mseri If the tagged release was indeed changed, should we instead just update the hash on the ocaml-repository?

Only if the content is unchanged. I am just about checking that

mseri commented:
:( opam.ocaml.org seems to be temporarily offline

mseri commented:
:( opam.ocaml.org seems to be temporarily offline

mseri commented:
I can confirm that they are unchanged, but since it is an autogenerated tag archive, which has proven to be unstable in the past, I have added the original zip to the opam-source-archive and pointed the sources to that one

R1kM commented:
Thanks!
